## Проект по SQL

## Описание данных <a id="2"></a>

`Таблица books`

Содержит данные о книгах:

* book_id — идентификатор книги;
* author_id — идентификатор автора;
* title — название книги;
* num_pages — количество страниц;
* publication_date — дата публикации книги;
* publisher_id — идентификатор издателя.

`Таблица authors`

Содержит данные об авторах:

* author_id — идентификатор автора;
* author — имя автора.

`Таблица publishers`

Содержит данные об издательствах:

* publisher_id — идентификатор издательства;
* publisher — название издательства;

`Таблица ratings`

Содержит данные о пользовательских оценках книг:

* rating_id — идентификатор оценки;
* book_id — идентификатор книги;
* username — имя пользователя, оставившего оценку;
* rating — оценка книги.

`Таблица reviews`

Содержит данные о пользовательских обзорах:

* review_id — идентификатор обзора;
* book_id — идентификатор книги;
* username — имя автора обзора;
* text — текст обзора.

## Описание проекта <a id="3"></a>

Компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

## Задания <a id="3"></a>

* Посчитайте, сколько книг вышло после 1 января 2000 года;
* Для каждой книги посчитайте количество обзоров и среднюю оценку;
* Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
* Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
* Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

## Цели исследования <a id="3"></a>

1. Определить количество книг, которые были выпущены после 1 января 2000 года, чтобы оценить современность и актуальность коллекции книг в базе данных.

2. Для каждой книги, содержащейся в базе данных, определить количество обзоров и среднюю оценку, чтобы оценить популярность книг и качество контента.

3. Определить издательство, которое выпустило наибольшее количество книг, толщина которых составляет 50 страниц или более, чтобы оценить наиболее активного и успешного издателя.

4. Определить автора с самой высокой средней оценкой книг, учитывая только книги с 50 и более оценками, чтобы оценить наиболее успешных и талантливых авторов.

5. Посчитать среднее количество обзоров от пользователей, которые поставили более 50 оценок, чтобы определить, насколько активны пользователи в базе данных, и какие книги наиболее популярны.

## Загрузка и просмотр данных

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
 'pwd': 'Sdf4$2;d-d30pp', # пароль
 'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
 'port': 6432, # порт подключения
 'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

In [2]:
query = '''
SELECT *
FROM books
LIMIT 10;
'''

sample_df = pd.read_sql_query(query, con=engine) 
sample_df

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   
5        6        257               1st to Die (Women's Murder Club  #1)   
6        7        258               2nd Chance (Women's Murder Club  #2)   
7        8        260              4th of July (Women's Murder Club  #4)   
8        9        563                                   A Beautiful Mind   
9       10        445                                 A Bend in the Road   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  
5        424       2005-05-20           116  
6        400       2005-05-20           116  
7        448       2006-06-01           318  
8        461       2002-02-04           104  
9        341       2005-04-01           116

In [3]:
query = '''
SELECT *
FROM authors
LIMIT 10;
'''

sample_df = pd.read_sql_query(query, con=engine) 
sample_df

author_id                                author
0          1                            A.S. Byatt
1          2        Aesop/Laura Harris/Laura Gibbs
2          3                       Agatha Christie
3          4                         Alan Brennert
4          5              Alan Moore/David   Lloyd
5          6                            Alan Paton
6          7           Albert Camus/Justin O'Brien
7          8                         Aldous Huxley
8          9    Aldous Huxley/Christopher Hitchens
9         10  Aleksandr Solzhenitsyn/H.T. Willetts

In [4]:
query = '''
SELECT *
FROM publishers
LIMIT 10;
'''

sample_df = pd.read_sql_query(query, con=engine) 
sample_df

publisher_id                                publisher
0             1                                      Ace
1             2                                 Ace Book
2             3                                Ace Books
3             4                            Ace Hardcover
4             5        Addison Wesley Publishing Company
5             6                                  Aladdin
6             7                       Aladdin Paperbacks
7             8                             Albin Michel
8             9                          Alfred A. Knopf
9            10  Alfred A. Knopf Books for Young Readers

In [5]:
query = '''
SELECT *
FROM ratings
LIMIT 10;
'''

sample_df = pd.read_sql_query(query, con=engine) 
sample_df

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2
5          6        3  johnsonamanda       4
6          7        3    scotttamara       5
7          8        3    lesliegibbs       5
8          9        4    abbottjames       5
9         10        4   valenciaanne       4

In [6]:
query = '''
SELECT *
FROM reviews
LIMIT 10;
'''

sample_df = pd.read_sql_query(query, con=engine) 
sample_df

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   
5          6        3    lesliegibbs   
6          7        4   valenciaanne   
7          8        4    abbottjames   
8          9        5        npowers   
9         10        5        staylor   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...  
5           Analysis no several cause international.  
6  One there cost another. Say type save. With pe...  
7  Within enough mother. There at system full rec...  
8  Thank now focus realize economy focus fly. Ite...  
9  Game push lot reduce where remember. Including...

## Считаем, сколько книг вышло после 1 января 2000 года

In [7]:
# напишите запрос
query_1 = '''
SELECT COUNT(*) AS total_books
FROM books
WHERE publication_date > '2000-01-01';
'''
task_1 = pd.read_sql_query(query_1, con=engine) 
# выполните запрос
task_1

total_books
0          819

### Вывод

После 1 января 2000 года вышло 819 книг.

## Для каждой книги посчитаем количество обзоров и среднюю оценку

In [8]:
# напишите запрос
query_2 = '''
SELECT 
  books.title, 
  COUNT(DISTINCT reviews.review_id) AS total_reviews, 
  AVG(ratings.rating) AS avg_rating
FROM 
  books
LEFT JOIN 
  ratings ON books.book_id = ratings.book_id
LEFT JOIN 
  reviews ON books.book_id = reviews.book_id
GROUP BY 
  books.book_id
ORDER BY avg_rating DESC, total_reviews DESC;
'''
task_2 = pd.read_sql_query(query_2, con=engine) 
# выполните запрос
task_2

title  total_reviews  \
0                        A Dirty Job (Grim Reaper  #1)              4   
1              School's Out—Forever (Maximum Ride  #2)              3   
2         Moneyball: The Art of Winning an Unfair Game              3   
3        Arrows of the Queen (Heralds of Valdemar  #1)              2   
4    Wherever You Go  There You Are: Mindfulness Me...              2   
..                                                 ...            ...   
995  The World Is Flat: A Brief History of the Twen...              3   
996                                      Drowning Ruth              3   
997                  His Excellency: George Washington              2   
998                                              Junky              2   
999                               Harvesting the Heart              2   

     avg_rating  
0          5.00  
1          5.00  
2          5.00  
3          5.00  
4          5.00  
..          ...  
995        2.25  
996        2.00  
997        2.00  
998        2.00  
999        1.50  

[1000 rows x 3 columns]

### Вывод

В данном выводе можно увидеть информацию по кол-ву обзоров и среднему рейтингу по каждой книге. Также данные отфильтрованы по среднему рейтингу и кол-ву обзоров в порядке убывания.

Наивысший рейтинг и большее кол-во обзоров у книги "A Dirty Job (Grim Reaper #1)", наименьший - "Harvesting the Heart".

## Определим издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

In [9]:
# напишите запрос
query_3 = '''
SELECT publishers.publisher, COUNT(books.book_id) as book_count
FROM books
JOIN publishers ON books.publisher_id = publishers.publisher_id
WHERE books.num_pages > 50
GROUP BY publishers.publisher
ORDER BY book_count DESC
LIMIT 1;
'''
task_3 = pd.read_sql_query(query_3, con=engine) 
# выполните запрос
task_3

publisher  book_count
0  Penguin Books          42

### Вывод

Больше всего книг, в которых больше 50 страниц, выпустило издательство Penguin Books - 42 книги.

## Определим автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [10]:
# напишите запрос
query_4 = '''
SELECT authors.author, AVG(ratings.rating) as avg_rating
FROM (
    SELECT book_id
    FROM ratings
    GROUP BY book_id
    HAVING COUNT(rating_id) > 50
) AS rated_books
JOIN books ON rated_books.book_id = books.book_id
JOIN authors ON books.author_id = authors.author_id
JOIN ratings ON books.book_id = ratings.book_id
GROUP BY authors.author
ORDER BY avg_rating DESC
LIMIT 1;
'''
task_4 = pd.read_sql_query(query_4, con=engine) 
# выполните запрос
task_4

author  avg_rating
0  J.K. Rowling/Mary GrandPré    4.287097

### Вывод

Автор с самой высокой средней оценкой 4.29 - J.K. Rowling/Mary GrandPré, среди книг с 50 и более оценками.

## Посчитаем среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [11]:
# напишите запрос
query_5 = '''
SELECT AVG(review_count) AS avg_review_count
FROM (
  SELECT COUNT(*) AS review_count
  FROM reviews
  WHERE username IN (
    SELECT username
    FROM ratings
    GROUP BY username
    HAVING COUNT(*) > 50
  )
  GROUP BY username
) AS counts;
'''
task_5 = pd.read_sql_query(query_5, con=engine) 
# выполните запрос
task_5

avg_review_count
0         24.333333

### Вывод

Среднее кол-во обзоров от пользователей, которые поставили больше 50 оценков - 24.33

## Общий вывод

* После 1 января 2000 года вышло 819 книг,
* Наивысший рейтинг и большее кол-во обзоров у книги "A Dirty Job (Grim Reaper #1)", наименьший - "Harvesting the Heart" (если рассматривать все книги),
* Больше всего книг, в которых больше 50 страниц, выпустило издательство Penguin Books - 42 книги,
* Автор с самой высокой средней оценкой 4.29 - J.K. Rowling/Mary GrandPré, среди книг с 50 и более оценками,
* Среднее кол-во обзоров от пользователей, которые поставили больше 50 оценков - 24.33.

Предложения для стартапа:
* Создать систему поощрения пользователей, которые оставляют много отзывов и оценок. Среднее количество обзоров от пользователей, которые поставили больше 50 оценок, достаточно высоко, и это может быть индикатором того, что они являются важными читателями и могут помочь в продвижении книг через свои рекомендации. Например, можно предложить пользователям скидки на подписку за каждые 10-20 оставленных отзывов,
* Сделать упор на книги, которые имеют высокий рейтинг и большее количество обзоров, например, J.K. Rowling. Пользователи склонны доверять мнению других читателей, и книги с большим количеством положительных отзывов могут привлекать больше внимания к подписке.